### PR TITLE
Add an abstraction for custom query caches

### PR DESCRIPTION
**Pull request by Zoxc**
r? @eddyb

### CODE REVIEW

**Comment by Zoxc** [comment]
@rust-timer build 8e818dc3304a24344d80a0481fee50f50da399a1

File: src/librustc/ty/query/keys.rs [hunk 1]
```diff
@@ -12,7 +13,9 @@ use rustc_span::{Span, DUMMY_SP};
 
 /// The `Key` trait controls what types can legally be used as the key
 /// for a query.
-pub(super) trait Key {
+pub trait Key {
+    type CacheSelector;
```
Comment (Centril) [inline src/librustc/ty/query/keys.rs]:
```suggestion
    type CacheSelector = DefaultCacheSelector;
```

**Comment by Zoxc** [comment]
Checking to see if removing the `storage` attributes gives a more neutral performance impact.

@bors try @rust-timer queue

**Comment by Zoxc** [comment]
I removed the changes which added and used a new query cache for some queries so that can be measured by perf separately. I did some performance tuning so now the [performance looks good](https://perf.rust-lang.org/compare.html?start=6317721cd951030f1f28801c34931c3c2322bc06&end=6e4fb02185fae88af7d0774391141e0f3bfca114).

File: src/librustc/query/mod.rs [hunk 2]
```diff
@@ -663,7 +665,8 @@ rustc_queries! {
         /// associated types. This is almost always what you want,
         /// unless you are doing MIR optimizations, in which case you
         /// might want to use `reveal_all()` method to change modes.
-        query param_env(_: DefId) -> ty::ParamEnv<'tcx> {}
+        query param_env(_: DefId) -> ty::ParamEnv<'tcx> {
+        }
```
Comment (eddyb) [inline src/librustc/query/mod.rs]: Are these automatic formatting changes, or can they be removed?

File: src/librustc/ty/query/keys.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc/ty/query/keys.rs]: @Zoxc Is this actionable?

**Review by eddyb — APPROVED** [review]
r=me modulo nits

File: src/librustc/ty/query/keys.rs (hunk 1 above)
Comment (Zoxc) [reply src/librustc/ty/query/keys.rs]: It's unstable and I ran into some issues trying to use it.

File: src/librustc/query/mod.rs (hunk 2 above)
Comment (Zoxc) [reply src/librustc/query/mod.rs]: No, this is just an incomplete reversion. I've rebased these away.

**Comment by Zoxc** [comment]
@bors r=eddyb p=1

cc @michaelwoerister

**Pull request merged**